### ISSUE TITLE
Issue with single-dimensional variables

### ISSUE THREAD

**Issue by Technici4n** [coin-or/CyLP#96]
Hello, this may have been reported before but there is an annoying issue with size one variables when trying to multiply them by a matrix.

Here is an example. I am encoding the constraints `1 <= x <= 2` as `x <= 2` and `-x <= -1`.
```python
from cylp.cy import CyClpSimplex
from cylp.py.modeling.CyLPModel import CyLPArray
import numpy as np

s = CyClpSimplex()
x = s.addVariable('x', 1)
A = np.matrix([[1, -1]])
b = np.array([2, -1])
s.addConstraint(A*x <= b)
s.objective = x.sum()
s.primal()
```

This example crashes with the following error message:
```
IndexError                                Traceback (most recent call last)
<ipython-input-3-43d30ca758a0> in <module>
      7 A = np.matrix([[1, -1]])
      8 b = np.array([2, -1])
----> 9 s.addConstraint(A*x <= b)
     10 s.objective = x.sum()
     11 s.primal()

cylp\cy\CyClpSimplex.pyx in cylp.cy.CyClpSimplex.CyClpSimplex.addConstraint()

C:\ProgramData\Anaconda3\lib\site-packages\cylp\py\modeling\CyLPModel.py in addConstraint(self, cons, consName, addMpsNames)
   1016         '''
   1017 
-> 1018         c = cons.evaluate(consName)
   1019         if not c.isRange:
   1020             self.constraints.append(c)

C:\ProgramData\Anaconda3\lib\site-packages\cylp\py\modeling\CyLPModel.py in evaluate(self, name)
    291                     right = operands.pop()
    292                     left = operands.pop()
--> 293                     cons.perform(token, left, right)
    294                     operands.append(CyLPExpr(token, left, right))
    295 

C:\ProgramData\Anaconda3\lib\site-packages\cylp\py\modeling\CyLPModel.py in perform(self, opr, left, right)
    402                             raise Exception("Coefficient:\n%s\n has %d" \
    403                                             " columns expected %d"
--> 404                                         % (left, left.shape[1], right.dim))
    405                         if len(left.shape) == 1:  # Array
    406                             nr = 1

IndexError: tuple index out of range
```

I think the issue comes from `np.squeeze`ing the matrix at the beginning of `CyLPConstraint.perform` ([link](https://github.com/coin-or/CyLP/blob/master/cylp/py/modeling/CyLPModel.py#L358)) which transforms the matrix to an array when one of the dimensions is 1.
I don't know if this can be fixed easily?
I ran into this issue when solving over multiple datasets with a loop: one of the datasets only had one number and created a size 1 variable.
I managed to work around the issue by adding an extra variable and padding my matrices with zeros. A bit annoying and ugly but it did the trick.

**Comment by tkralphs** [comment]
I guess I don't understand exactly what the problem you are trying to solve in the above example looks like mathematically. In general, the shape of `A` has to match with the shape of `x` just for mathematical correctness, putting aside possible bugs in CyLP. You have a matrix `A` with shape `(1,2)`, so for `A*x` to be interpretable mathematically, `x` must have shape `(2,1)`. You can't multiple a vector of variables that has shape `(1, 1)` with a matrix that has shape `(1,2)`. 

I guess maybe you are thinking that since you can multiply a whole matrix by a scalar (which is sort of like a matrix with shape `(1, 1)`), then this should also work with a one-dimensional variable, but it can't work that way for a range of reasons. If that's what you want, then it is equivalent to just sum the columns of A (and the objective coefficients) to get a matrix with shape `(*,1)`, which could then be multiplied with a single-dimensional variable. 

So I guess this error seems to be expected, given your example.

**Comment by Technici4n** [comment]
My bad, the example is wrong, sorry about that.
`A` should have shape `(2, 1)` and I'm trying to multiply it by `x` with shape `(1, 1)` to get a `(2, 1)` result. This is the corrected code, which gives the same error even though I modified `A`:
```python
from cylp.cy import CyClpSimplex
from cylp.py.modeling.CyLPModel import CyLPArray
import numpy as np

s = CyClpSimplex()
x = s.addVariable('x', 1)
A = np.matrix([[1], [-1]])
b = np.array([2, -1])
s.addConstraint(A*x <= b)
s.objective = x.sum()
s.primal()
```

**Comment by tkralphs** [comment]
Ah, OK, that makes more sense. I will try to have a look at this, it may be related to some other pending issues.

**Comment by juhaszad** [comment]
Hi, I have also ran into this issue, were you able to solve it?

**Comment by Technici4n** [comment]
You can work around it by using a length 2 vector and only use the first component, that's what I ended up doing!

**Comment by juhaszad** [comment]
Yeah, thank you, I did the same, I just thought whether there were some 'in-built' trick in cylp which solves the issue. Thanks again! :)
